File: Prácticas/P4/JAVA/Civitas/src/civitas/CivitasJuego.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;
import java.util.*;


/**
 *
 * @author Jose Padial
 */
public class CivitasJuego {
    
    private int indiceJugadorActual;
    Dado dado = Dado.getInstance();
    OperacionesJuego operacion;
    MazoSorpresas mazo;
    Tablero tablero;
    ArrayList<Jugador> jugadores = new ArrayList<>();
    EstadosJuego estado;
    GestorEstados gestorEstados;
    public Diario diario = Diario.getInstance();
    
 
    CivitasJuego(ArrayList<String> nombres){
          
        for (String i :nombres){
            jugadores.add(new Jugador(i) {});
        }
        
        gestorEstados = new GestorEstados();
        estado = gestorEstados.estadoInicial();
        
        indiceJugadorActual = 0; // Debo implementar un generador de numeros aleatorios
        
        mazo = new MazoSorpresas();
        
        inicializarTablero(mazo);
        inicializarMazoSorpresas(tablero);
    }
    
    private void avanzaJugador(){
        Jugador jugadorActual = jugadores.get(indiceJugadorActual);
        int posAct = jugadorActual.getNumCasillaActual();
        int tirada = dado.tirar();
        int posNew = tablero.nuevaPosicion(posAct, tirada);
        Casilla casillaActual = tablero.getCasilla(posNew);
        contabilizarPasosPorSalida(jugadorActual);
        jugadorActual.moverACasilla(posNew);
        casillaActual.recibeJugador(indiceJugadorActual, jugadores);
        contabilizarPasosPorSalida(jugadorActual);
    }
    
    public boolean cancelarHipoteca(int ip){
        return jugadores.get(indiceJugadorActual).cancelarHipoteca(ip);
    }
    
    public boolean comprar(){
        int numCas = jugadores.get(indiceJugadorActual).getNumCasillaActual();
        TituloPropiedad propiedad = ((CasillaCalle)tablero.getCasilla(numCas)).getTituloPropiedad();
                
        return jugadores.get(indiceJugadorActual).comprar(propiedad);
    }
    
  public boolean construirCasa(int ip){
        return jugadores.get(indiceJugadorActual).construirCasa(ip);
    }
    
    public boolean construirHotel(int ip){
        return jugadores.get(indiceJugadorActual).construirHotel(ip);
    }
    
    private void contabilizarPasosPorSalida(Jugador jugadorActual){
        if (tablero.getPorSalida() > 0){
            jugadorActual.pasaPorSalida();
        }
    }
    
    public boolean finalDelJuego(){
        boolean bancarrota = false;
        
        for (Jugador i :jugadores){
            if(i.enBancarrota() == true){
                bancarrota = true;
            }
        }
        
        return bancarrota;
    }
    
    public Casilla getCasillaActual(){
        int numCas = jugadores.get(indiceJugadorActual).getNumCasillaActual();
        return tablero.getCasilla(numCas);
    }
    
    public Jugador getJugadorActual(){
        return jugadores.get(indiceJugadorActual);
    }
    
    public boolean hipotecar(int ip){
        return jugadores.get(indiceJugadorActual).hipotecar(ip);
    }
    
    public String infoJugadorTexto(){
        return jugadores.get(indiceJugadorActual).toString();
    }
    
    private void inicializarMazoSorpresas(Tablero tablero){
        mazo.alMazo(new SorpresaIrCarcel(tablero));
        mazo.alMazo(new SorpresaIrCasilla(tablero,10,"Desplazate hacia la casilla indicada"));
        mazo.alMazo(new SorpresaIrCasilla(tablero,16,"Desplazate hacia la casilla indicada"));
        mazo.alMazo(new SorpresaIrCasilla(tablero,tablero.getCarcel(),"A la carcel"));
        mazo.alMazo(new SorpresaPagarCobrar(tablero,1000,"Recibes la cantidad indicada"));
        mazo.alMazo(new SorpresaPagarCobrar(tablero,-1000,"Pierdes la cantidad indicada"));
        mazo.alMazo(new SorpresaPorCasaHotel(tablero,-50,"Debes pagar los impuestos de tus propiedades"));
        mazo.alMazo(new SorpresaPorCasaHotel(tablero,50,"Hacienda te ha hecho un regalito"));
        mazo.alMazo(new SorpresaPorJugador(tablero,-100,"El jugador debe pagar la cantidad indicada a cada jugador"));
        mazo.alMazo(new SorpresaPorJugador(tablero,100,"El jugador recibe la cantidad indicada de cada jugador"));
        mazo.alMazo(new SorpresaSalirCarcel(mazo));
        
    }
    
    private void inicializarTablero(MazoSorpresas mazo){
        tablero = new Tablero(8);
        tablero.aniadeCasilla(new CasillaCalle(new TituloPropiedad("Iglesia", 100, (float)19.0, 300, 150, 300)));
        tablero.aniadeCasilla(new CasillaCalle(new TituloPropiedad("C/ Istán", 75, (float)19.0, 100, 100, 500)));
        tablero.aniadeCasilla(new CasillaCalle(new TituloPropiedad("C/ Don Vito", 125, (float)19.0, 200, 200, 350)));
        tablero.aniadeCasilla(new CasillaSorpresa("SORPRESA",mazo));
        tablero.aniadeCasilla(new CasillaCalle(new TituloPropiedad("C/ Pepe Osorio", 50, (float)19.0, 100, 300, 350)));
        tablero.aniadeCasilla(new CasillaCalle(new TituloPropiedad("C/ Pajaritos", 150, (float)19.0, 300, 500, 300)));
        tablero.aniadeJuez();
        tablero.aniadeCasilla(new CasillaSorpresa("SORPRESA",mazo));
        tablero.aniadeCasilla(new CasillaCalle(new TituloPropiedad("C/ Lagasca", 200, (float)19.0, 175, 120, 450)));
        tablero.aniadeCasilla(new CasillaCalle(new TituloPropiedad("C/ San Miguel", 75, (float)19.0, 225, 150, 225)));
        tablero.aniadeCasilla(new CasillaImpuesto("IMPUESTO",75));
        tablero.aniadeCasilla(new CasillaCalle(new TituloPropiedad("C/ Pizarro", 215, (float)19.0, 420, 350, 200)));
        tablero.aniadeCasilla(new CasillaSorpresa("SORPRESA",mazo));
        tablero.aniadeCasilla(new CasillaCalle(new TituloPropiedad("C/ Oriental", 140, (float)19.0, 330, 500, 380)));
        tablero.aniadeCasilla(new Casilla("PARKING"));
        tablero.aniadeCasilla(new CasillaCalle(new TituloPropiedad("C/ Dolores", 200, (float)19.0, 400, 550, 325)));
        tablero.aniadeCasilla(new CasillaCalle(new TituloPropiedad("C/ Caballero", 300, (float)19.0, 400, 700, 500)));
        tablero.aniadeCasilla(new CasillaCalle(new TituloPropiedad("C/ Ruiz", 300, (float)19.0, 500, 800, 550)));
        
    }
    
    public void pasarTurno(){
        indiceJugadorActual = (indiceJugadorActual + 1) % jugadores.size();
    }
    
    public ArrayList<Jugador> ranking(){
        ArrayList<Jugador> ranking;
        ranking = jugadores;
        Collections.sort(ranking);
        return ranking;
    }
    
    public boolean salirCarcelPagando(){
        return jugadores.get(indiceJugadorActual).salirCarcelPagando();
    }
    
    public boolean salirCarcelTirando(){
        return jugadores.get(indiceJugadorActual).salirCarcelTirando();
    }
    
    public OperacionesJuego siguientePaso(){
        Jugador jugadorActual = jugadores.get(indiceJugadorActual);
        operacion = gestorEstados.operacionesPermitidas(jugadorActual,estado);
        
        if(operacion == OperacionesJuego.PASAR_TURNO){
            pasarTurno();
            siguientePasoCompletado(operacion);
        }
        else if(operacion == OperacionesJuego.AVANZAR){
            avanzaJugador();
            siguientePasoCompletado(operacion);
        }
        return operacion;
    }
    
    public void siguientePasoCompletado(OperacionesJuego operacion){
        estado = gestorEstados.siguienteEstado(jugadores.get(indiceJugadorActual), estado, operacion);
    }
    
    public boolean vender(int ip){
       return jugadores.get(indiceJugadorActual).vender(ip);
    }
    
}

File: Prácticas/P4/RUBY/Civitas/lib/civitas_juego.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative"mazo_sorpresas.rb"
require_relative"tablero.rb"
require_relative"operaciones_juego.rb"
require_relative"dado.rb"
require_relative"jugador.rb"
require_relative"estados_juego.rb"
require_relative"gestor_estados.rb"
require_relative"diario.rb"
require_relative"casilla.rb"
require_relative"casilla_calle.rb"
require_relative"casilla_impuesto.rb"
require_relative"casilla_juez.rb"
require_relative"casilla_sorpresa.rb"
require_relative"controlador.rb"
require_relative"operacion_inmobiliaria.rb"
require_relative"operaciones_inmobiliarias.rb"
require_relative"respuestas.rb"
require_relative"sorpresa_convertirse.rb"
require_relative"sorpresa_ir_carcel.rb"
require_relative"sorpresa_ir_casilla.rb"
require_relative"sorpresa_pagar_cobrar.rb"
require_relative"sorpresa_por_casa_hotel.rb"
require_relative"sorpresa_por_jugador.rb"
require_relative"sorpresa_salida_carcel.rb"
require_relative"salidas_carcel.rb"
require_relative"titulo_propiedad.rb"
require_relative"vista_textual.rb"

module Civitas
class Civitas_juego
  
  public
  def initialize(nombres)
    @indice_jugador_actual 
    @dado = Dado.instance
    @operacion
    @mazo = Mazo_sorpresas.new(false)
    @tablero = Tablero.new(8)
    @jugadores = Array.new
    @gestor_estados = Gestor_estados.new
    @estado = @gestor_estados.estado_inicial
    @diario = Diario.instance
   
    for i in nombres
      @jugadores << Jugador.new(i)
    end
    
    @indice_jugador_actual = @dado.tirar % @jugadores.length
        
    inicializar_tablero(@mazo);
    inicializar_mazo_sorpresa(@tablero);
  end
  
  def cancelar_hipoteca(ip)
    return @jugadores.at(@indice_jugador_actual).cancelar_hipoteca(ip)
  end
  
  def comprar
     numCas = @jugadores.at(@indice_jugador_actual).get_num_casilla_actual
     propiedad = @tablero.get_casilla(numCas).titulo
                
     return @jugadores.at(@indice_jugador_actual).comprar(propiedad)
  end
  
  def construir_casa(ip)
    @jugadores.at(@indice_jugador_actual).construir_casa(ip)
  end
  
  def construir_hotel(ip)
    @jugadores.at(@indice_jugador_actual).construir_hotel(ip)
    puts "HE ENTRADO"
  end
  
  def final_del_juego
    final = false
    tamanio = @jugadores.length
    i = 0
    
    while(i<tamanio) 
      if(@jugadores.at(i).en_bancarrota == true)
        final = true
      end
      i = i+1
    end
    
    return final
    
  end
  
  def get_casilla_actual
    numCas = @jugadores.at(@indice_jugador_actual).get_num_casilla_actual
    return @tablero.get_casilla(numCas)
  end
  
  def get_diario
    return @diario
  end
  
  def get_jugador_actual
    return @jugadores.at(@indice_jugador_actual)
  end
  
  def hipotecar(ip)
    return @jugadores.at(@indice_jugador_actual).hipotecar(ip)
  end
  
  def info_jugador_texto
    return @jugadores.at(@indice_jugador_actual).to_s
  end
  
  def salir_carcel_pagando
    return @jugadores.at(@indice_jugador_actual).salir_carcel_pagando
  end
  
  def salir_carcel_tirando
    return @jugadores.at(@indice_jugador_actual).salir_carcel_tirando
  end
  
  def siguiente_paso_completado(operacion)
    @estado = @gestor_estados.siguiente_estado(@jugadores.at(@indice_jugador_actual), @estado, operacion);
  end
  
  def vender(ip)
    @jugadores.at(@indice_jugador_actual).vender(ip)
  end
  
  def ranking
    ranking = Array.new
    ranking = @jugadores
    ranking.sort
    return ranking
  end
  
  def avanza_jugador
    jugadorActual = @jugadores.at(@indice_jugador_actual)
    posAct = jugadorActual.get_num_casilla_actual
    tirada = @dado.tirar
    posNew = @tablero.nueva_posicion(posAct,tirada)
    casillaActual = @tablero.get_casilla(posNew)
    contabilizar_pasos_por_salida(jugadorActual)
    jugadorActual.mover_a_casilla(posNew)
    casillaActual.recibe_jugador(@indice_jugador_actual, @jugadores)
    contabilizar_pasos_por_salida(jugadorActual)
  end
  
  def contabilizar_pasos_por_salida(jugador_actual)
    if(@tablero.get_por_salida > 0)
        jugador_actual.pasa_por_salida
    end
  end
  
  def inicializar_mazo_sorpresa(tablero)
    @mazo.al_mazo(SorpresaIrCarcel.new(tablero,"Vas a la carcel"))
    @mazo.al_mazo(SorpresaIrCasilla.new(tablero,10,"Desplazate hacia la casilla indicada"))
    @mazo.al_mazo(SorpresaIrCasilla.new(tablero,16,"Desplazate hacia la casilla indicada"))
    @mazo.al_mazo(SorpresaIrCasilla.new(tablero,tablero.get_carcel,"A la carcel"))
    @mazo.al_mazo(SorpresaPagarCobrar.new(tablero,1000,"Recibes la cantidad indicada"))
    @mazo.al_mazo(SorpresaPagarCobrar.new(tablero,-1000,"Pierdes la cantidad indicada"))
    @mazo.al_mazo(SorpresaPorCasaHotel.new(tablero,-50,"Debes pagar los impuestos de tus propiedades"))
    @mazo.al_mazo(SorpresaPorCasaHotel.new(tablero,50,"Hacienda te ha hecho un regalito"))
    @mazo.al_mazo(SorpresaPorJugador.new(tablero,-100,"El jugador debe pagar la cantidad indicada a cada jugador"))
    @mazo.al_mazo(SorpresaPorJugador.new(tablero,100,"El jugador recibe la cantidad indicada de cada jugador"))
    @mazo.al_mazo(SorpresaSalidaCarcel.new(@mazo))
  end
  
  def inicializar_tablero(mazo)
    @tablero.aniade_casilla(CasillaCalle.new(Titulo_propiedad.new("Iglesia", 100, 19, 300, 150, 300)))
    @tablero.aniade_casilla(CasillaCalle.new(Titulo_propiedad.new("C/ Istan", 75, 19, 100, 100, 500)))
    @tablero.aniade_casilla(CasillaCalle.new(Titulo_propiedad.new("C/ Don Vito", 125, 19, 200, 200, 350)))
    @tablero.aniade_casilla(Casilla_Sorpresa.new("SORPRESA",mazo))
    @tablero.aniade_casilla(CasillaCalle.new(Titulo_propiedad.new("C/ Pepe Osorio", 50, 19, 100, 300, 350)))
    @tablero.aniade_casilla(CasillaCalle.new(Titulo_propiedad.new("C/ Pajaritos", 150, 19, 300, 500, 300)))
    @tablero.aniade_juez()
    @tablero.aniade_casilla(Casilla_Sorpresa.new("SORPRESA",mazo))
    @tablero.aniade_casilla(CasillaCalle.new(Titulo_propiedad.new("C/ Lagasca", 200, 19, 175, 120, 450)))
    @tablero.aniade_casilla(CasillaCalle.new(Titulo_propiedad.new("C/ San Miguel", 75, 19, 225, 150, 225)))
    @tablero.aniade_casilla(CasillaImpuesto.new("IMPUESTO",75))
    @tablero.aniade_casilla(CasillaCalle.new(Titulo_propiedad.new("C/ Pizarro", 215, 19, 420, 350, 200)))
    @tablero.aniade_casilla(Casilla_Sorpresa.new("SORPRESA",mazo))
    @tablero.aniade_casilla(CasillaCalle.new(Titulo_propiedad.new("C/ Oriental", 140, 19, 330, 500, 380)))
    @tablero.aniade_casilla(Casilla.new("PARKING"))
    @tablero.aniade_casilla(CasillaCalle.new(Titulo_propiedad.new("C/ Dolores", 200, 19, 400, 550, 325)))
    @tablero.aniade_casilla(CasillaCalle.new(Titulo_propiedad.new("C/ Caballero", 300, 19, 400, 700, 500)))
    @tablero.aniade_casilla(CasillaCalle.new(Titulo_propiedad.new("C/ Ruiz", 300, 19, 500, 800, 550)))
  end
  
  def pasar_turno
    @indice_jugador_actual = (@indice_jugador_actual + 1) % @jugadores.length
  end
  
  def siguiente_paso
    jugadorActual = @jugadores.at(@indice_jugador_actual)
    operacion = @gestor_estados.operaciones_permitidas(jugadorActual,@estado)
    puts "OPERACION PERMITIDA Y ESTADO " + operacion.to_s + " " + @estado.to_s
        
    if(operacion == Operaciones_juego::PASAR_TURNO)
            pasar_turno
            siguiente_paso_completado(operacion)
    end
         
    if(operacion == Operaciones_juego::AVANZAR)
            avanza_jugador
            siguiente_paso_completado(operacion)
    end
    
        return operacion
  end
  
end
end
